Add Querier to Extern

ethanfrey opened the pull request:
Closes #100 

Also add to the whole stack (including between vm and wasm).
I have updated all generic usages of Extern and tests pass again.

TODO:
- [x] Add query_chain to the context
- [x] Use external Querier from sample contract and test it

webmaster128 commented on contracts/hackatom/src/contract.rs:
```diff
@@ -200,6 +226,30 @@ mod tests {
+            address: rich_addr.clone(),
+        };
+        let query_response = query(&deps, query_msg).unwrap();
+        let bal: BalanceResponse = from_slice(query_response.as_slice()).unwrap();
```
it would be nice to have a `from_binary` to decode `Binary` directly

webmaster128 commented on packages/std/src/api.rs:
```diff
@@ -123,6 +123,7 @@ impl From<Error> for ApiError {
 pub enum ApiSystemError {
     InvalidRequest { source: String },
     NoSuchContract { addr: HumanAddr },
+    UnknownError {},
```
Symbol name can be simplified to `ApiSystemError::Unknown`

webmaster128 commented on packages/std/src/errors.rs:
```diff
@@ -81,6 +81,8 @@ pub enum SystemError {
         backtrace: snafu::Backtrace,
     },
+    #[snafu(display("Unknown system error"))]
+    UnknownError { backtrace: snafu::Backtrace },
```
Same here: `SystemError::Unknown`

webmaster128 commented on packages/std/src/imports.rs:
```diff
@@ -13,7 +16,10 @@ use crate::types::{CanonicalAddr, HumanAddr};
 
 // this should be plenty for any address representation
-static ADDR_BUFFER: usize = 72;
+static ADDR_BUFFER: usize = 100;
```
There is a global max length of `90` specified for bech32 strings: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32 So HRP+separator+data combined must ensure to not exceed this. This would be a good limit and source the decision here.

webmaster128 commented on packages/vm/src/compatability.rs:
```diff
@@ -93,7 +94,7 @@ fn find_missing_export(module: &Module, required_exports: &[&str]) -> Option<Str
         Some(export_section) => Vec::from(export_section.entries())
             .iter()
-            .map(|entry| format!("{}", entry.field()))
+            .map(|entry| entry.field().to_string())
```
🏅

webmaster128 commented on packages/vm/src/instance.rs:
```diff
@@ -107,6 +109,12 @@ where
                 }),
+                "humanize_address" => Func::new(move |ctx: &mut Ctx, canonical_ptr: u32, human_ptr: u32| -> i32 {
+                    do_human_address(api, ctx, canonical_ptr, human_ptr)
+                }),
```
The humanize_address was duplicated. Merge error?

ethanfrey replied:
Yup

webmaster128 commented on packages/vm/src/context.rs:
```diff
@@ -205,68 +245,97 @@ mod iter_support {
-struct ContextData<S: Storage> {
+struct ContextData<S: Storage, Q: Querier> {
     data: Option<S>,
+    querier: Option<Q>,
```
Is this really needed here? It pollutes all the storage handling. Isn't Querier more like Api than like Storage?

ethanfrey replied:
This is the big question.

To act like api, I need to implement Copy and Clone.
There is state there behind it.

However,  &Querier (non mutable) is Copy and Cloneable.

I could try a bigger refactor there, but this would probably touch all usages of Extern and I would rather try that after this PR.

ethanfrey replied:
Look at the constructor types:

```rust
impl<S, A, Q> Instance<S, A, Q>
where
    S: Storage + 'static,
    A: Api + 'static,
    Q: Querier + 'static,
{
    pub fn from_code(code: &[u8], deps: Extern<S, A, Q>, gas_limit: u64) -> Result<Self> {
```

I don't see a way to get `&Q` to implement `'static`

webmaster128 replied:
So this state behind `Querier` comes from the mock implementation, right?

```rust
/// A stateless convenience wrapper around imports provided by the VM
#[derive(Copy, Clone)]
pub struct ExternalQuerier {}
```

```rust
/// MockQuerier holds an immutable table of bank balances
/// TODO: also allow querying contracts
#[derive(Clone)]
pub struct MockQuerier {
    balances: HashMap<HumanAddr, Vec<Coin>>,
}
```

I know that it is a mayor annoyance to try using something without `Copy` in the import implementations. I tried to remove Copy from the Api trait once and failed horribly.

I see two ways here:
1. Find a way to implement the data in `MockQuerier` in a `Copy`able fashion
2. Find a way to clone api and querier into import implementations

I prefer 2., as it keeps us more flexible. But using the same overhead for Querier as the non-Clonable Storage seems not very elegant.

webmaster128 replied:
I remember the problem with cloning into import implementations: The callbacks need to be `Fn`, not `FnOnce`, i.e. they can be called multiple times. When you move a previously cloned value into the callback, it only exists for one function execution. And you cannot use a reference, since there is the function lives longer than the objects of the that are passed in.

Api only works because it can be copied without any lifetime restrictions.

The only place for long living data is `Ctx.data`, so the querier needs to be in `ContextData`.

But there is a simpler way to temporarily access the querier without all this move in, move out action:

```rust
pub fn with_querier_from_context<S: Storage, Q: Querier, F: FnMut(&Q)>(ctx: &Ctx, mut func: F) {
    let context_data = unsafe { Box::from_raw(ctx.data as *mut ContextData<S, Q>) };
    if let Some(q) = &context_data.querier {
        func(q);
    }
    Box::leak(context_data);
}
```

ethanfrey replied:
The problem with this last one is a bug you found a while ago.
What if this callback calls `assert!()` and fails (which is the same as `panic!()`)

In this case, it unwinds the stack and first drops the local variables, (incl `q` and `context_data`), then goes further up the stack and calls the destructor (`destroy unmanaged storage`). Which will free out context_data again.

My approach was to then leave context_data full of `None` while we had the data in local variables (so they only got cleared once on panic) and then reset them when finishing. I think you made a test for this, but we should make more.

ethanfrey replied:
> The only place for long living data is Ctx.data, so the querier needs to be in ContextData.

Exactly. If there is no data (or it is copyable), there is no problem.

webmaster128 commented on packages/std/src/mock.rs:
```diff
@@ -110,29 +123,34 @@ pub fn mock_env<T: Api, U: Into<HumanAddr>>(
 impl MockQuerier {
-    pub fn new(balances: HashMap<HumanAddr, Vec<Coin>>) -> Self {
-        MockQuerier { balances }
+    pub fn new(balances: Vec<(HumanAddr, Vec<Coin>)>) -> Self {
```
How do you feel about passing this as a read only reference (is slice of references) and then cloning it into the internal data store? I think it would simplify a lot of callers that don't need to clone at call time.

webmaster128 commented on packages/vm/src/context.rs:
```diff
@@ -110,6 +114,40 @@ pub fn do_human_address<A: Api>(api: A, ctx: &mut Ctx, canonical_ptr: u32, human
+    request_ptr: u32,
+    response_ptr: u32,
+) -> i32 {
+    let request = match read_region(ctx, request_ptr, MAX_LENGTH_ADDRESS) {
```
Query requests should get their own max length constant

ethanfrey commented on contracts/hackatom/src/contract.rs:
```diff
@@ -146,12 +158,26 @@ fn query_verifier<S: Storage, A: Api>(deps: &Extern<S, A>) -> Result<QueryRespon
+        Ok(Ok(res)) => Ok(res),
+    }?;
+    // in theory we would process the response, but here it is the same type, so just pass through
+    Ok(response)
```
Yes.

There are two levels of Result (this is not that unusual in rust world). But now I get it. I will adjust

ethanfrey replied:
I looked again.
I did what you said, but it was tricky with the `}?;` at the end, so Err() in match returned errors, and the Ok was unwrapped.
That is odd and confusing. Fixed the syntax.

ethanfrey commented:
Rebased on #222 and now addressing most comments

ethanfrey commented:
@webmaster128 Okay, I think I addressed all your comments.

The issue of using Querier like Api has to do with State (Api is stateless) and Copy. I would prefer to make a second PR/issue to see if that can be simplified.

Other than that, can you take another pass on this after rebase and cleanup?

ethanfrey commented on packages/vm/src/context.rs:
```diff
@@ -18,10 +22,12 @@ pub use iter_support::{
+use crate::serde::to_vec;
 
-static MAX_LENGTH_DB_KEY: usize = 100_000;
+static MAX_LENGTH_DB_KEY: usize = 2_000;
```
I will change it back. I didn't read that issue before

ethanfrey commented:
Doing one last rebase and cleanup before merge.

Follow up issue to dig into the ContextData stuff is here: https://github.com/CosmWasm/cosmwasm/issues/225

Lyla-Fischer commented:
FYI - This issue is still listed as open on works-hub.

ethanfrey commented:
Thanks for letting us know.
I just manually closed a number of those issues that were linked to closed github issues.
